core/src/main/java/com/scalar/db/transaction/consensuscommit/PrepareMutationComposer.java

```diff
@@ -137,28 +160,35 @@ private void add(Get base) {
     mutations.add(put);
   }
 
-  private List<Value<?>> createBeforeValues(Mutation base, TransactionResult result) {
-    List<Value<?>> values = new ArrayList<>();
+  private List<Column<?>> createBeforeColumns(Mutation base, TransactionResult result) {
+    List<Column<?>> columns = new ArrayList<>();
     result
-        .getValues()
+        .getColumns()
         .values()
         .forEach(
-            v -> {
-              if (isBeforeRequired(v, base.getPartitionKey(), base.getClusteringKey())) {
-                values.add(v.copyWith(Attribute.BEFORE_PREFIX + v.getName()));
+            column -> {
+              if (isBeforeRequired(column, base.getPartitionKey(), base.getClusteringKey())) {
+                if (column.getName().equals(Attribute.VERSION) && column.hasNullValue()) {
+                  // For preparing an update of a deemed-commit state record, we need to use
+                  // version 0 rather than NULL since we want to distinguish the preparation from
+                  // an initial record insertion.
+                  columns.add(IntColumn.of(Attribute.BEFORE_VERSION, 0));
+                } else {
+                  columns.add(column.copyWith(Attribute.BEFORE_PREFIX + column.getName()));
+                }
               }
             });
-    return values;
+    return columns;
   }
 
-  private boolean isBeforeRequired(Value<?> value, Key primary, Optional<Key> clustering) {
-    return !value.getName().startsWith(Attribute.BEFORE_PREFIX)
-        && !isValueInKeys(value, primary, clustering);
+  private boolean isBeforeRequired(Column<?> column, Key primary, Optional<Key> clustering) {
+    return !column.getName().startsWith(Attribute.BEFORE_PREFIX)
```

Comment:
This might be a silly question, but what if an original column name starts with `before_` prefix like `before_update_balance` (non-primary or clustering key)? This logic seems to return false, but I think `before_before_update_balance` column should be taken care of.

Reply:
Oh, good point. Yeah, it should be taken care of **if we allow users to use a column name starting with `before_`.** For new projects, at least, prohibiting such column names would not cause serious problems. For existing databases, some users might be using such column names, but I guess it would be rare, and we could ask users to rename them. So, IMHO, we can prohibit user columns with `before_` in both cases for now.

@brfrn169 By the way, I think we don't have any guard for those names in the schema-loader, and I was able to create a `before_xx` column and faced an issue that its **before** column was not updated in the prepare phase. Is my understanding correct?

Reply:
Ah, I can reproduce it with the current code.

**ScalarDB SQL CLI**
```
0: scalardb> select * from example_app.account2;
+----+----------+-----+
| id | before_x |  y  |
+----+----------+-----+
| 1  | 10       | 100 |
+----+----------+-----+
1 row selected (0.005 seconds)
0: scalardb> update example_app.account2 set before_x = 20, y = 200 where id = 1;
1 row affected (0.04 seconds)
```

**psql**
```
postgres=# select id, before_x, y, tx_state, before_before_x, before_y from example_app.account2;
 id | before_x |  y  | tx_state | before_before_x | before_y 
----+----------+-----+----------+-----------------+----------
  1 |       20 | 200 |        3 |                 |      100
(1 row)
```

Reply:
@komamitsu Thanks! Yeah, I faced the same issue.

Reply:
I thought we accept to create the `before_xx` column as a normal column, and in this case, its metadata column should be `before_before_xx`. But, we might miss something to handle that. 

The logic to detect the before column is here:
https://github.com/scalar-labs/scalardb/blob/master/core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java#L179-L204

The basic idea is, if there are both `xxx` and `before_xx` (before prefixed) columns in the table, `xxx` is an after image column, and `before_xx` is a before image column.

Reply:
> By the way, I think we don't have any guard for those names in the schema-loader,

Ah, yes, it looks like we don't have any guard for that. I think we should add some validation for that in `ConsensusCommitAdmin.createTable()`:
https://github.com/scalar-labs/scalardb/blob/44a7baf62937b9a44e19d36516b20dcdd7130264/core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitAdmin.java#L83-L88

Reply:
> Ah, yes, it looks like we don't have any guard for that. I think we should add some validation for that in ConsensusCommitAdmin.createTable():

@brfrn169 @komamitsu Thank you for the comment. So, let's fix it in another PR.

Reply:
We can probably think of introducing configurable prefix?
(e.g., if `before_` is already used, we set `prev_` as the prefix for before columns)
It's gonna be one time configuration before starting though.

Reply:
> configurable prefix

Sounds reasonable to me! Though it depends on the balance of the benefit and engineering side effort (Analytics also needs to know the configuration.)

BTW, it's a bit off-topic, but I noticed `metadata` table doesn't have any information on whether it's a `before` column. I think it would be useful for `before` column check in a robust way. Maybe adding a new type `BEFORE` to the existing `key_type`...?

Reply:
```java
       return tableMetadata 
           .getColumnNames() 
           .contains(columnName.substring(Attribute.BEFORE_PREFIX.length())); 
```
https://github.com/scalar-labs/scalardb/pull/841#discussion_r1171961632

I also think this logic works for most cases except that a user has used both `x` and `before_x` in their database and wants to migrate the table to ScalarDB.

Reply:
@jnmt Sorry, just a confirmation. In the current logic, does creating the `before_xx` column as a normal column work or not? I think it works.

> I think we should add some validation for that in `ConsensusCommitAdmin.createTable()`:

I meant here that we should add the validation to prevent from being created both `xxx` and `before_xxx` as normal columns. I think just creating a `before_xxx` column works currently. Is there a bug in this?

Reply:
> > I think we should add some validation for that in `ConsensusCommitAdmin.createTable()`:
> 
> I meant here that we should add the validation to prevent from being created both `xxx` and `before_xxx` as normal columns. I think just creating a `before_xxx` column works currently. Is there a bug in this?

@brfrn169 Oh, that's my misunderstanding, but anyway, I don't think it works well even in the current code, according to reproductions by Mitsu(https://github.com/scalar-labs/scalardb/pull/841#discussion_r1171065262) and me. Probably a bug unfortunately.

Reply:
Ah, okay. It should be a bug. Let me check. Thanks.

Reply:
Ah, okay we should use `ConsensusCommitUtils.isAfterImageColumn()` here:

https://github.com/scalar-labs/scalardb/blob/44a7baf62937b9a44e19d36516b20dcdd7130264/core/src/main/java/com/scalar/db/transaction/consensuscommit/PrepareMutationComposer.java#L155

I thought I had fixed it 😞

Reply:
Oh, we already have the validation to check if a table has transaction meta columns when creating the table: 
https://github.com/scalar-labs/scalardb/blob/44a7baf62937b9a44e19d36516b20dcdd7130264/core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java#L46-L48

Reply:
@brfrn169 Thanks for the additional info. But I guess it only checks to prevent from being created both xxx and before_xxx as normal columns. We can still create before_xxx column itself...

Reply:
@jnmt Yes, the following bug fix allows users to create `before_` prefixed columns:

https://github.com/scalar-labs/scalardb/pull/844